packages/opamp-client-node/lib/opamp-client.js

```diff
@@ -0,0 +1,864 @@
+                },
+                'ignore _scheduleSend'
+            );
+            return;
+        }
+
+        // Schedule the send in `delayMs`, unless one is already scheduled for
+        // sooner or `overrideExisting`.
+        const sendTime = Date.now() + delayMs;
+        if (
+            overrideExisting ||
+            !this._nextSendTime ||
+            sendTime < this._nextSendTime
+        ) {
+            this._log.trace({delayMs}, 'schedule next opamp send');
+            if (this._nextSendTimeout) {
+                clearTimeout(this._nextSendTimeout);
+            }
+            this._nextSendTime = sendTime;
+            this._nextSendTimeout = setTimeout(() => {
+                this._nextSendTime = null;
+                this._sendMsg();
+            }, delayMs);
+            this._nextSendTimeout.unref();
+            if (this._diagChs && !this._shutdown) {
+                this._diagChs[DIAG_CH_SEND_SCHEDULE].publish({
+                    instanceUidStr: this._instanceUidStr,
+                    name: DIAG_CH_SEND_SCHEDULE,
+                    time: Date.now(),
+                    delayMs,
+                    errCount: this._numSendFailures,
+                });
+            }
+        }
+    }
+
+    /**
+     * Send a status update to the server, then schedule the next send.
+     * Data to send is either on `this._queue`, or this is a simple heartbeat.
+     *
+     * Dev Note: This sets `this._sending = true` for its duration. This
+     * function cannot throw, otherwise the client will be wedged.
+     *
+     * Error handling behaviour is loosely derived from the OpAMP spec,
+     * including sections "Establishing Connection", "Retrying Messages" and
+     * "Throttling". It isn't always clear the exact intent:
+     *
+     * - if `client.request()` rejects (no conn, headersTimeout): log.error, re-enqueue, schedule with backoff(30s .. 5min)
+     * - if HTTP 429 with valid `Retry-After` header: log.debug, re-enqueue, schedule for given time (min 30s)
+     * - if HTTP 503 with valid `Retry-After` header: log.error, re-enqueue, schedule for given time (min 30s)
+     * - if HTTP != 200 || content-type != protobuf: log.error, re-enqueue, schedule with backoff(30s .. 5min)
+     * - if `await res.body.bytes()` rejection (bodyTimeout): log.error, re-enqueue, schedule with backoff(30s .. 5min)
+     * - if s2a.errorResponse:
+     *     - if .type == "Unavailable" && .Details.retryAfterNanoseconds > 0: log.debug, re-enqueue, schedule for given time (min 30s)
+     *     - if .type == "BadRequest": log.error, do *not* re-enqueue, schedule for given time (min 30s)
+     *     - else: log.error, re-enqueue, schedule with backoff(30s .. 5min)
+     * - else success: process the `s2a` message
+     */
+    async _sendMsg() {
+        this._sending = true;
+        this._log.trace({queue: this._queue}, '_sendMsg start');
+
+        /** @type {AgentToServer} */
+        let msg = {
```

Comment:
Also to avoid type issues. this is a bit convoluted 
```suggestion
        /** @type {Parameters<typeof import('@bufbuild/protobuf').create<typeof AgentToServerSchema>>[1]} */
        const msg = {
```

I think it would be better to have it as a separate type along with other type definitions

```js
/**
 * @typedef {Parameters<typeof import('@bufbuild/protobuf').create<typeof AgentToServerSchema>>[1]} AgentToServerPayload
 */
/**
 * @typedef {Parameters<typeof import('@bufbuild/protobuf').create<typeof ServerToAgentSchema>>[1]} ServerToAgentPayload
 */

```